cmd/bee-join/main.go

```diff
@@ -0,0 +1,158 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/spf13/cobra"
+)
+
+var (
+	outFilePath  string // flag variable, output file
+	outFileForce bool   // flag variable, overwrite output file if exists
+	host         string // flag variable, http api host
+	port         int    // flag variable, http api port
+	ssl          bool   // flag variable, uses https for api if set
+)
+
+// apiStore provies a storage.Getter that retrieves chunks from swarm through the HTTP chunk API.
+type apiStore struct {
+	baseUrl string
+}
+
+// newApiStore creates a new apiStore
+func newApiStore(host string, port int, ssl bool) (storage.Getter, error) {
+	scheme := "http"
+	if ssl {
+		scheme += "s"
+	}
+	u := &url.URL{
+		Host:   fmt.Sprintf("%s:%d", host, port),
+		Scheme: scheme,
+		Path:   "bzz-chunk",
+	}
+	return &apiStore{
+		baseUrl: u.String(),
+	}, nil
+}
+
+// Get implements storage.Getter
+func (a *apiStore) Get(ctx context.Context, mode storage.ModeGet, address swarm.Address) (ch swarm.Chunk, err error) {
+	addressHex := address.String()
+	url := strings.Join([]string{a.baseUrl, addressHex}, "/")
+	res, err := http.DefaultClient.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("chunk %s not found", addressHex)
+	}
+	chunkData, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	ch = swarm.NewChunk(address, chunkData)
+	return ch, nil
+}
+
+// Join is the underlying procedure for the CLI command
+func Join(cmd *cobra.Command, args []string) (err error) {
+	// if output file is specified, create it if it does not exist
+	var outFile *os.File
+	if outFilePath != "" {
+		// make sure we have full path
+		outDir := filepath.Dir(outFilePath)
+		if outDir != "." {
+			err := os.MkdirAll(outDir, 0o777) // skipcq: GSC-G301
+			if err != nil {
+				return err
+			}
+		}
+		// protect any existing file unless explicitly told not to
+		outFileFlags := os.O_CREATE | os.O_WRONLY
+		if outFileForce {
+			outFileFlags |= os.O_TRUNC
+		} else {
+			outFileFlags |= os.O_EXCL
+		}
+		// open the file
+		outFile, err = os.OpenFile(outFilePath, outFileFlags, 0o666) // skipcq: GSC-G302
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+	} else {
+		outFile = os.Stdout
+	}
+
+	// process the reference to retrieve
+	addr, err := swarm.ParseHexAddress(args[0])
+	if err != nil {
+		return err
+	}
+
+	// initialize interface with HTTP API
+	store, err := newApiStore(host, port, ssl)
+	if err != nil {
+		return err
+	}
+
+	// create the join and get its data reader
+	j := joiner.NewSimpleJoiner(store)
+	r, l, err := j.Join(context.Background(), addr)
+	if err != nil {
+		return err
+	}
+
+	// join, rinse, repeat until done
+	teeReader := io.TeeReader(r, outFile)
```

Comment:
i kind of don't understand why do you need a `TeeReader` here. Isn't `tee` needed when you need to split the output into two different streams? but here it appears you read into two streams indeed, but one of them is never used (i.e. you never use `data`)

Reply:
> not sure if I see the benefit of having a tee here

it does exactly what you're describing, only without handling it directly.

Reply:
i think `Copy` won't work because the joiner needs to read at whole chunk-size intervals. but then in that case you could just `Read` as you do here and after the read explicitly `Write` to the `Writer`. `TeeReader` is still not needed. Please read the source code for what TeeReader does on Read

Reply:
>> not sure if I see the benefit of having a tee here

> it does exactly what you're describing, only without handling it directly.

what's your argument exactly @acud? I don't get it.

Reply:
my argument is that you're reading in to the byte slice just in order to get the tee reader to `Write` into the writer, nothing else. then why don't you call `Write` explicitly from where you `Read` in the loop below and get rid of the tee reader?

Reply:
Understood your point after a call.